File: sdk/typescript/src/middleware.ts
import type {
  AccessPolicy,
  JsonValue,
  ObserveRequest,
  ObserveResponse,
  RecallRequest,
  RecallResponse,
  RecallTrace,
  RequestContext,
  RequestOptions,
} from "./types.js";

export interface MemoryClient {
  observe(request: ObserveRequest, options?: RequestOptions): Promise<ObserveResponse>;
  recall(request: RecallRequest, options?: RequestOptions): Promise<RecallResponse>;
}

export interface AgentSessionOptions {
  readonly context: RequestContext;
  readonly access: AccessPolicy;
  readonly agentId: string;
  readonly sessionId: string;
}

export interface BeforeTurnOptions extends RequestOptions {
  readonly pageSize?: number;
  readonly atCommit?: number | null;
  readonly continuation?: string | null;
}

export interface AfterTurnOptions extends RequestOptions {
  readonly metadata?: Readonly<Record<string, JsonValue>>;
  readonly idempotencyKey?: string;
}

async function turnKey(
  sessionId: string,
  agentId: string,
  sequence: number,
  userMessage: string,
  assistantResponse: string,
): Promise<string> {
  if (globalThis.crypto?.subtle === undefined) {
    throw new Error("Web Crypto SHA-256 is required for deterministic idempotency keys");
  }
  const encoder = new TextEncoder();
  const chunks: Uint8Array[] = [encoder.encode("contextdb-agent-turn-v1\0")];
  const addText = (value: string): void => {
    const bytes = encoder.encode(value);
    const length = new Uint8Array(8);
    new DataView(length.buffer).setBigUint64(0, BigInt(bytes.byteLength), false);
    chunks.push(length, bytes);
  };
  addText(sessionId);
  addText(agentId);
  const encodedSequence = new Uint8Array(8);
  new DataView(encodedSequence.buffer).setBigUint64(0, BigInt(sequence), false);
  chunks.push(encodedSequence);
  addText(userMessage);
  addText(assistantResponse);
  const size = chunks.reduce((sum, chunk) => sum + chunk.byteLength, 0);
  const encoded = new Uint8Array(size);
  let offset = 0;
  for (const chunk of chunks) {
    encoded.set(chunk, offset);
    offset += chunk.byteLength;
  }
  const digest = new Uint8Array(await globalThis.crypto.subtle.digest("SHA-256", encoded));
  const hex = Array.from(digest, (item) => item.toString(16).padStart(2, "0")).join("");
  return `agent-session:${sessionId}:${sequence}:${hex}`;
}

// AgentSession is provider-neutral before/after-turn orchestration. It keeps
// RequestContext explicit and never treats it as authentication.
export class AgentSession {
  private readonly client: MemoryClient;
  private readonly context: RequestContext;
  private readonly access: AccessPolicy;
  private readonly agentId: string;
  private readonly sessionId: string;
  private sequenceValue = 0;
  private recallSequence = 0;
  private lastTraceValue: RecallTrace | null = null;
  private lastContinuationValue: string | null = null;
  private tail: Promise<void> = Promise.resolve();

  constructor(client: MemoryClient, options: AgentSessionOptions) {
    if (options.agentId === "" || options.sessionId === "") {
      throw new TypeError("agentId and sessionId must be non-empty");
    }
    if (options.context.workspace_id !== options.access.workspace_id) {
      throw new TypeError("access policy and request context workspace differ");
    }
    this.client = client;
    this.context = {
      ...options.context,
      audiences: [...options.context.audiences],
      scopes: [...options.context.scopes],
    };
    this.access = {
      ...options.access,
      scopes: [...options.access.scopes],
      owners: [...options.access.owners],
      audience: [...options.access.audience],
      audience_purpose_grants: Object.fromEntries(
        Object.entries(options.access.audience_purpose_grants).map(([audience, purposes]) => [
          audience,
          [...purposes],
        ]),
      ),
      purposes: [...options.access.purposes],
    };
    this.agentId = options.agentId;
    this.sessionId = options.sessionId;
  }

  get sequence(): number {
    return this.sequenceValue;
  }

  get lastTrace(): RecallTrace | null {
    return this.lastTraceValue;
  }

  get lastContinuation(): string | null {
    return this.lastContinuationValue;
  }

  beforeTurn(message: string, options: BeforeTurnOptions = {}): Promise<RecallResponse> {
    return this.exclusive(async () => {
      const request: RecallRequest = {
        context: {
          ...this.context,
          request_id: `session:${this.sessionId}:before:${this.recallSequence}`,
        },
        query: message,
        page_size: options.pageSize ?? 20,
        at_commit: options.atCommit ?? null,
        continuation: options.continuation ?? null,
      };
      const requestOptions: { signal?: AbortSignal } = {};
      if (options.signal !== undefined) requestOptions.signal = options.signal;
      const response = await this.client.recall(request, requestOptions);
      this.recallSequence += 1;
      this.lastTraceValue = response.trace;
      this.lastContinuationValue = response.continuation;
      return response;
    });
  }

  afterTurn(userMessage: string, assistantResponse: string, options: AfterTurnOptions = {}): Promise<ObserveResponse> {
    return this.exclusive(async () => {
      const sequence = this.sequenceValue;
      const content = {
        kind: "chat_turn",
        session_id: this.sessionId,
        agent_id: this.agentId,
        sequence,
        user_message: userMessage,
        assistant_response: assistantResponse,
      } satisfies JsonValue;
      const metadata = {
        ...(options.metadata ?? {}),
        kind: "agent_session_turn",
        session_id: this.sessionId,
        agent_id: this.agentId,
        sequence,
      } satisfies Record<string, JsonValue>;
      const request: ObserveRequest = {
        context: {
          ...this.context,
          request_id: `session:${this.sessionId}:after:${sequence}`,
        },
        idempotency_key:
          options.idempotencyKey ||
          (await turnKey(this.sessionId, this.agentId, sequence, userMessage, assistantResponse)),
        observation_id: `session:${this.sessionId}:turn:${sequence}`,
        metadata,
        content,
        access: this.access,
      };
      const requestOptions: { signal?: AbortSignal } = {};
      if (options.signal !== undefined) requestOptions.signal = options.signal;
      const response = await this.client.observe(request, requestOptions);
      this.sequenceValue += 1;
      return response;
    });
  }

  private async exclusive<T>(operation: () => Promise<T>): Promise<T> {
    const previous = this.tail;
    let release: (() => void) | undefined;
    this.tail = new Promise<void>((resolve) => {
      release = resolve;
    });
    await previous;
    try {
      return await operation();
    } finally {
      release?.();
    }
  }
}

File: sdk/go/middleware.go
package contextdb

import (
	"context"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"fmt"
	"sync"
)

// MemoryClient is the narrow seam required by agent-session middleware.
type MemoryClient interface {
	Observe(context.Context, ObserveRequest) (ObserveResponse, error)
	Recall(context.Context, RecallRequest) (RecallResponse, error)
}

type BeforeTurnOptions struct {
	PageSize     uint32
	AtCommit     *uint64
	Continuation *string
}

type AfterTurnOptions struct {
	Metadata       map[string]any
	IdempotencyKey string
}

// AgentSession serializes turn numbering so retries retain the same
// observation identity and idempotency key.
type AgentSession struct {
	client    MemoryClient
	context   RequestContext
	access    AccessPolicy
	agentID   string
	sessionID string

	mu               sync.Mutex
	sequence         uint64
	recallSequence   uint64
	lastTrace        *RecallTrace
	lastContinuation *string
}

func NewAgentSession(
	client MemoryClient,
	requestContext RequestContext,
	access AccessPolicy,
	agentID string,
	sessionID string,
) (*AgentSession, error) {
	if client == nil {
		return nil, errors.New("client is required")
	}
	if agentID == "" || sessionID == "" {
		return nil, errors.New("agent ID and session ID must be non-empty")
	}
	if requestContext.WorkspaceID != access.WorkspaceID {
		return nil, errors.New("access policy and request context workspace differ")
	}
	return &AgentSession{
		client:    client,
		context:   cloneRequestContext(requestContext),
		access:    cloneAccessPolicy(access),
		agentID:   agentID,
		sessionID: sessionID,
	}, nil
}

func cloneStrings(values []string) []string {
	return append([]string(nil), values...)
}

func cloneRequestContext(value RequestContext) RequestContext {
	value.Audiences = cloneStrings(value.Audiences)
	value.Scopes = cloneStrings(value.Scopes)
	return value
}

func cloneAccessPolicy(value AccessPolicy) AccessPolicy {
	value.Scopes = cloneStrings(value.Scopes)
	value.Owners = cloneStrings(value.Owners)
	value.Audience = cloneStrings(value.Audience)
	value.Purposes = cloneStrings(value.Purposes)
	grants := make(map[string][]string, len(value.AudiencePurposeGrants))
	for audience, purposes := range value.AudiencePurposeGrants {
		grants[audience] = cloneStrings(purposes)
	}
	value.AudiencePurposeGrants = grants
	return value
}

func (session *AgentSession) BeforeTurn(
	ctx context.Context,
	message string,
	options BeforeTurnOptions,
) (RecallResponse, error) {
	session.mu.Lock()
	defer session.mu.Unlock()
	pageSize := options.PageSize
	if pageSize == 0 {
		pageSize = 20
	}
	requestContext := session.context
	requestContext.RequestID = fmt.Sprintf(
		"session:%s:before:%d",
		session.sessionID,
		session.recallSequence,
	)
	response, err := session.client.Recall(ctx, RecallRequest{
		Context:      requestContext,
		Query:        message,
		PageSize:     pageSize,
		AtCommit:     options.AtCommit,
		Continuation: options.Continuation,
	})
	if err != nil {
		return RecallResponse{}, err
	}
	session.recallSequence++
	trace := response.Trace
	session.lastTrace = &trace
	if response.Continuation == nil {
		session.lastContinuation = nil
	} else {
		continuation := *response.Continuation
		session.lastContinuation = &continuation
	}
	return response, nil
}

func (session *AgentSession) AfterTurn(
	ctx context.Context,
	userMessage string,
	assistantResponse string,
	options AfterTurnOptions,
) (ObserveResponse, error) {
	session.mu.Lock()
	defer session.mu.Unlock()
	content := map[string]any{
		"kind":               "chat_turn",
		"session_id":         session.sessionID,
		"agent_id":           session.agentID,
		"sequence":           session.sequence,
		"user_message":       userMessage,
		"assistant_response": assistantResponse,
	}
	metadata := make(map[string]any, len(options.Metadata)+4)
	for key, value := range options.Metadata {
		metadata[key] = value
	}
	metadata["kind"] = "agent_session_turn"
	metadata["session_id"] = session.sessionID
	metadata["agent_id"] = session.agentID
	metadata["sequence"] = session.sequence
	idempotencyKey := options.IdempotencyKey
	if idempotencyKey == "" {
		digest := turnDigest(
			session.sessionID,
			session.agentID,
			session.sequence,
			userMessage,
			assistantResponse,
		)
		idempotencyKey = fmt.Sprintf(
			"agent-session:%s:%d:%s",
			session.sessionID,
			session.sequence,
			hex.EncodeToString(digest[:]),
		)
	}
	requestContext := session.context
	requestContext.RequestID = fmt.Sprintf(
		"session:%s:after:%d",
		session.sessionID,
		session.sequence,
	)
	response, err := session.client.Observe(ctx, ObserveRequest{
		Context:        requestContext,
		IdempotencyKey: idempotencyKey,
		ObservationID: fmt.Sprintf(
			"session:%s:turn:%d",
			session.sessionID,
			session.sequence,
		),
		Metadata: metadata,
		Content:  content,
		Access:   session.access,
	})
	if err != nil {
		return ObserveResponse{}, err
	}
	session.sequence++
	return response, nil
}

func turnDigest(
	sessionID string,
	agentID string,
	sequence uint64,
	userMessage string,
	assistantResponse string,
) [sha256.Size]byte {
	hash := sha256.New()
	_, _ = hash.Write([]byte("contextdb-agent-turn-v1\x00"))
	writeText := func(value string) {
		var length [8]byte
		binary.BigEndian.PutUint64(length[:], uint64(len([]byte(value))))
		_, _ = hash.Write(length[:])
		_, _ = hash.Write([]byte(value))
	}
	writeText(sessionID)
	writeText(agentID)
	var encodedSequence [8]byte
	binary.BigEndian.PutUint64(encodedSequence[:], sequence)
	_, _ = hash.Write(encodedSequence[:])
	writeText(userMessage)
	writeText(assistantResponse)
	var result [sha256.Size]byte
	copy(result[:], hash.Sum(nil))
	return result
}

func (session *AgentSession) LastTrace() (RecallTrace, bool) {
	session.mu.Lock()
	defer session.mu.Unlock()
	if session.lastTrace == nil {
		return RecallTrace{}, false
	}
	return *session.lastTrace, true
}

func (session *AgentSession) LastContinuation() (string, bool) {
	session.mu.Lock()
	defer session.mu.Unlock()
	if session.lastContinuation == nil {
		return "", false
	}
	return *session.lastContinuation, true
}

func (session *AgentSession) Sequence() uint64 {
	session.mu.Lock()
	defer session.mu.Unlock()
	return session.sequence
}
